theano/tensor/opt.py

```diff
@@ -3052,9 +3122,52 @@ def local_remove_switch_const_cond(node):
                 out = T.alloc(out, *[node.outputs[0].shape[i] for i
                                      in xrange(out.ndim)])
             return [out]
-
+        # if left is right -> left
+        if node.inputs[1] is node.inputs[2]:
+            return [node.inputs[1]]
+        # if left equal right -> left
+        if (T.extract_constant(node.inputs[1]) ==
+            T.extract_constant(node.inputs[2])):
+            if node.inputs[1].type == node.outputs[0].type:
+                return [node.inputs[1]]
+            if node.inputs[2].type == node.outputs[0].type:
+                return [node.inputs[2]]
+        # This case happen with scan.
+        # Elemwise{switch}(le(shape_i{id}(X), 0), 0, shape_i{id}(X)) -> shape_i{id}(X)
+        left = node.inputs[1]
+        right = node.inputs[2]
+        if (cond.owner and
+            isinstance(cond.owner.op, T.Elemwise) and
+            isinstance(cond.owner.op.scalar_op, scalar.LE) and
+            cond.owner.inputs[0].owner and
+            isinstance(cond.owner.inputs[0].owner.op, Shape_i) and
+            T.extract_constant(cond.owner.inputs[1]) == 0 and
+            T.extract_constant(left) == 0 and
+            right is cond.owner.inputs[0]):
+            assert right.type == node.outputs[0].type
+            return [right]
         return False
     return False
+local_remove_switch_const_cond = local_useless_switch
+
+
+#@register_canonicalize
+#@register_specialize
+@gof.local_optimizer([Shape_i])
+def local_shape_i_infered(node):
+    if not isinstance(node.op, Shape_i):
+        return
+    if not hasattr(node, 'fgraph'):
+        return
+    if not hasattr(node.fgraph, 'shape_feature'):
+        return
+    try:
+        shp = node.fgraph.shape_feature.shape_of[node.inputs[0]][node.op.i]
+        c = get_scalar_constant_value(shp)
+        import pdb;pdb.set_trace()
+        return [T.constant(c, dtype=node.outputs[0].dtype)]
+    except NotScalarConstantError:
```

Comment:
why this opt is needed? I don't see why this is linked to the current opt you are working on. I'm not sure this is a good idea. Did I forgot something?